### PR TITLE
Batched multi class classification

### PR DESCRIPTION
**Pull request by yoavkatz**
This PR adds and an example  explores the effect of long context in classification.  
It converts a standard multi class classification dataset (sst2 sentiment classification),
where single sentence texts are classified one by one, to a dataset that
where multiple sentences are classified using  a single LLM call.  
It compares the f1_micro in both approaches on two models.
It uses serializers to verbalize and enumerated list of multiple sentences and labels.

It added a new operator called CollateInstances.

### CODE REVIEW

**Comment by yoavkatz** [comment]
Example model input for batch_size=10:
```
    <|begin_of_text|><|start_header_id|>system<|end_header_id|>
    
    <|eot_id|><|start_header_id|>user<|end_header_id|>
    
    Classify each of the texts to its corresponding sentiment from one of these options:
    1. positive
    2. negative
    Return for each index the correspond class in a separate line.
    Texts:
    1. comes from the brave , uninhibited performances 
    2. excruciatingly unfunny and pitifully unromantic 
    3. enriched by an imaginatively mixed cast of antic spirits 
    4. which half of dragonfly is worse : the part where nothing 's happening , or the part where something 's happening 
    5. in world cinema 
    6. very good viewing alternative 
    7. the plot is nothing but boilerplate clichés from start to finish , 
    8. the action is stilted 
    9. on all cylinders 
    10. will find little of interest in this film , which is often preachy and poorly acted <|eot_id|><|start_header_id|>assistant<|end_header_id|>
    
    Answer:
    1. positive
    2. negative
    3. positive
    4. negative
    5. positive
    6. positive
    7. negative
    8. negative
    9. positive
    10. negative<|eot_id|><|start_header_id|>user<|end_header_id|>
    
    Classify each of the texts to its corresponding sentiment from one of these options:
    1. positive
    2. negative
    Return for each index the correspond class in a separate line.
    Texts:
    1. it 's a charming and often affecting journey . 
    2. unflinchingly bleak and desperate 
    3. allows us to hope that nolan is poised to embark a major career as a commercial yet inventive filmmaker . 
    4. the acting , costumes , music , cinematography and sound are all astounding given the production 's austere locales . 
    5. it 's slow -- very , very slow . 
    6. although laced with humor and a few fanciful touches , the film is a refreshingly serious look at young women . 
    7. a sometimes tedious film . 
    8. or doing last year 's taxes with your ex-wife . 
    9. you do n't have to know about music to appreciate the film 's easygoing blend of comedy and romance . 
    10. in exactly 89 minutes , most of which passed as slowly as if i 'd been sitting naked on an igloo , formula 51 sank from quirky to jerky to utter turkey . <|eot
    _id|><|start_header_id|>assistant<|end_header_id|>
    
    Answer:
```

Seems llama and granite models are robust  to batching.

|    | model                              |   batch_size |   num_instances |   f1_micro |   ci_low |   ci_high |
|---:|:-----------------------------------|-------------:|----------------:|-----------:|---------:|----------:|
|  0 | ibm/granite-8b-instruct-preview-4k |           50 |              18 |       0.94 |     0.93 |      0.96 |
|  1 | ibm/granite-8b-instruct-preview-4k |           30 |              30 |       0.95 |     0.93 |      0.97 |
|  2 | ibm/granite-8b-instruct-preview-4k |           10 |              88 |       0.95 |     0.93 |      0.96 |
|  3 | ibm/granite-8b-instruct-preview-4k |            1 |             872 |       0.92 |     0.91 |      0.94 |
|  4 | meta-llama/llama-3-8b-instruct     |          100 |               9 |       0.87 |     0.82 |      0.9  |
|  5 | meta-llama/llama-3-8b-instruct     |           50 |              18 |       0.9  |     0.88 |      0.93 |
|  6 | meta-llama/llama-3-8b-instruct     |           10 |              88 |       0.89 |     0.87 |      0.91 |
|  7 | meta-llama/llama-3-8b-instruct     |            1 |             872 |       0.91 |     0.88 |      0.92 |

**Review by elronbandel — APPROVED** [review]
Thats very cool and can make a nice blog post!

File: examples/evaluate_batched_multiclass_classification.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            card=card,
+            template_card_index=0,
+            format=format,
+            num_demos=1,
+            demos_pool_size=5,
+            loader_limit=10000,
+            max_test_instances=1000 / batch_size,
+        )
+
+        test_dataset = dataset["test"]
```
Comment (elronbandel) [inline examples/evaluate_batched_multiclass_classification.py:122]: More efficient:
```suggestion
        test_dataset = load_dataset(
            card=card,
            template_card_index=0,
            format=format,
            num_demos=1,
            demos_pool_size=5,
            loader_limit=10000,
            max_test_instances=1000 / batch_size,
        )
```

Comment (yoavkatz) [reply examples/evaluate_batched_multiclass_classification.py:122]: This did not work.

File: examples/evaluate_batched_multiclass_classification.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        from unitxt.inference import IbmGenAiInferenceEngine
+
+        inference_model = IbmGenAiInferenceEngine(
+            model_name=model_name, max_new_tokens=1024
+        )
```
[the remark this answers is not in the document]
Comment (yoavkatz) [reply examples/evaluate_batched_multiclass_classification.py]: Done.

**Pull request merged**